Do not reset customization in FO when changing combination

zuk3975 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/8/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Solves issue https://github.com/PrestaShop/PrestaShop/issues/29185. Altough its a quick solution, some backend validation of id_customization (when customization is required) is still missing
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Resolves #29185, resolves #28041
| Related PRs       | 
| How to test?      | refer to #29185. :warning: Requires recompiling front core assets to take effect (`./tools/assets/build.sh front-core`)
| Possible impacts? | I don't know if reseting customization_id was important, but for example now if you switch combination and don't add another customization data, you will have previous one's values (which seems to be normal to me, but maybe im wrong?)


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

zuk3975 committed:
do not reset customization in FO when changing combination

zuk3975 commented:
@PrestaShop/committers 
just a quckfix in front, but backend validation is still missing.
I think maybe i could check if product requires a customization and depending on that throw error in backend CartController::init(). Altough current implementation is so unstable, so i don't see another option. Its probably possible that CartController will not always have id_product set in request, so in such cases we would probably juts need to ignore this validation as in further steps it would still fail right?

zuk3975 commented:
btw @sLorenzini I manage to reproduce it in 8.0.x also (and i haven't tested, but I think it should also be the same in 178, because the related code seems to be the same across these versions)

mparvazi commented:
@zuk3975 

After product has added to Cart, the last `product_customization_id` still remains on form.

zuk3975 committed:
refill customizationId value when updating quantity, but reset it elseway

zuk3975 commented:
updated.
So It used to reset the id_customization value on every update action (including the quantity update, which also caused adding product without customization to cart).
So I added condition to check if event is updatedProductQuantity and depending on that reset the id_customization or refill it with previous value.
It is still only the frontend "validation", my previous question about validating backend is still valid, so im waiting for suggestions. Also If anyone think of any other impact my changes might have done? (for now it seems to work as expected, but as legacy code is so mixed I wouldn't be surprized if I broke something :thinking:)

mparvazi commented on themes/_core/js/product.js:
```diff
@@ -228,7 +228,14 @@ function updateProduct(event, eventType, updateUrl) {
-        $(prestashop.selectors.product.inputCustomization).val(0);
+
+        // refill customizationId input value when updating quantity
+        if (eventType === 'updatedProductQuantity' && data.id_customization) {
```
I think also need to check `updatedProductCombination`.

![Screenshot from 2022-08-03 17-20-24](https://user-images.githubusercontent.com/85633460/182612054-10102b56-d16a-493d-84ac-093234f726e8.png)

zuk3975 committed:
do not reset id_customization when changing combination too

Hlavtox commented:
Duplicate of https://github.com/PrestaShop/PrestaShop/pull/29182, but this solution seems better.

The reset is there to fix an issue when you added the customized product to cart, but the ID of customization stayed in this input. So, when you added another piece to cart, it added the customized one, instead of "clear" one.
https://github.com/PrestaShop/PrestaShop/pull/22145

HanaRebaiQA commented:
Hello 

Thanks @Hlavtox for your catch.
ping @PrestaShop/prestashop-core-developers Which PR will be kept to fix the [issue](https://github.com/PrestaShop/PrestaShop/issues/29185) ? is't the first one created [PR1](https://github.com/PrestaShop/PrestaShop/pull/29182) ? or the [current](https://github.com/PrestaShop/PrestaShop/pull/29228) one ?

Thank You!

mparvazi commented:
I didn't get what's the problem about this PR?

matks commented:
@HanaRebaiQA I trust @Hlavtox and I think this PR might be the best 😉

mparvazi commented:
> > I didn't get what's the problem about this PR?
> 
> I think QA team needs us to choose what PR to keep 😉

Definitely this PR.

HanaRebaiQA commented:
Thank You @matks and @mparvazi 

So, @matks this [issue](https://github.com/PrestaShop/PrestaShop/pull/29182) needs to be closed ?

jolelievre reviewed (commented):
The fix seems legit, I'm just not sure about the target branch because the issue is not clear about where the bug happened first, we'll see this with QA team to make sure of the target.

NeOMakinG reviewed (dismissed):
🍬

atomiix commented:
Hi @PrestaShop/qa-functional, it looks like this PR solve #29228 as well, can you have a look?

hibatallahAouadni commented:
Hello @atomiix 

> Hi https://github.com/orgs/PrestaShop/teams/qa-functional, it looks like this PR solve https://github.com/PrestaShop/PrestaShop/pull/29228 as well, can you have a look?

I see that you mistyped the `issue_id` :sweat_smile: you put this `PR_id` :wink: 

Thanks!

atomiix commented:
@hibatallahAouadni you're right 🤦‍♂️, I updated my comment!

hibatallahAouadni commented:
@atomiix I add it to the Fixed tickets :wink:

florine2623 reviewed (commented):
Hello @zuk3975 ,

Thanks for the PR!

Launched automated tests ✅ 

The bug #29185 is fixed.
Tested the following:
- switch on/off to require a customisation in BO
- modify custom, save, modify again, etc
- delete/edit quantity from cart
- check product customisation popup in cart

Combination is never duplicated or hidden in the FO if existing ✅ 

I cannot check for #28041, because I can't view the shopping cart.
Here are the steps to reproduce:
- add custom product to cart
- go to BO > Orders > Shopping Carts
- click latest cart available: OK ✅ 
- in BO > Catalog > Products, modify custom product from Standard to Combination product
- add all sizes available > generate combination
- open FO in incognito mode to generate new cart
- add your modified custom product > add customization > Choose size > add to cart
- go back to BO > Orders > Shopping Carts
- click latest cart available: can’t open it, NOK :x:
- go to another BO page, example Orders, see this error x amount of time you clicked on the cart :x:

<img width="1226" alt="Screenshot 2022-09-08 at 16 41 29" src="https://user-images.githubusercontent.com/16019289/189152942-3b1b7d7f-5435-41bd-aa0e-b6a426009de5.png">

https://user-images.githubusercontent.com/16019289/189153183-4eabe233-2741-4fd1-8ea8-f7c7781976ec.mov

Could you check please ?
Thanks!

MatShir commented:
@zuk3975 Seems we have some QA feedback, we can use this opportunity to switch of branch. wdyt?

zuk3975 committed:
fix loading cart for viewing when it contains product with recently changed type (standard->combination).

zuk3975 commented:
closing in favour of
https://github.com/PrestaShop/PrestaShop/pull/29635
Cherry picked the commits and targeted 178 branch (couldn't rebase the same PR for some git issues which idk how to solve)
